### PR TITLE
Fix #70 projectiles issues

### LINKED ISSUE #70: Projectiles not properly exploding when bCutReplicationAfterSpawn=false
Projectiles are left hanging for 200ms after exploding on surfaces.
In multiplayer projectiles go through players and continue flying for 200ms after dealing damage.

Caused by client code trying to Destroy() from client-side which is not allowed on networked actors (unless bNetTemporary=true).

### PR DESCRIPTION
**Pull request by ChatouilleGG**
When projectiles have `bNetTemporary=true`, they only replicate spawning, but replication link is cut right away.
This is the simplest case, client is fully responsible for exploding/destroying projectile on its own.
This is suited for all spammable, less important projectiles.
It can be set in Blueprints through the `bCutReplicationAfterSpawn` property.

This mechanism is however not suited for projectiles that may need to be updated from server, such as interactive projectiles, or simply the more prominent projectiles where we want a more accurate simulation.
In that case, when projectile explodes on server, we replicate explosion info to correct client's simulation if necessary. This has the following purposes :
- Projectile may be detonated by user (combo or C4 type firemode), this cannot be predicted by other clients simulations.
- Projectile may hit something on server which it did not hit on client. Replication of explosion will correct that very quickly instead of letting projectile continue its flight on client machine (and generate unreg later on).
- Projectile may hit on client but not on server (typical unreg case), which we can detect due to the lack of explosion replication.

This replication of explosion leads to 4 potential cases to handle when dealing with client-side explosion :

**(1) Explode on client before receiving server explosion**
--> Play explosion, hide projectile, and wait for server confirmation.
--> We will either receive server explosion soon (confirm case), or much later (unreg case).

**(2) Explode on client after receiving server explosion**
--> This should not happen if we properly disable simulation in case (3).

**(3) Receive server explosion before exploded on client**
--> Play explosion, stop simulating and destroy when possible.
(note: client cannot destroy as long as replication link exists)

**(4) Receive server explosion after exploded on client**
--> As per case (1), this is either a confirmation, or an unreg case.
--> In unreg case, we should replay the real explosion at the proper location.

### CODE REVIEW

**Comment by wail** [comment]
I believe this PR should be to OpenTournament:unstable not OpenTournament:master

File: Source/OpenTournament/UR_Projectile.cpp [hunk 1]
```diff
@@ -209,24 +210,80 @@ void AUR_Projectile::DealSplashDamage()
...
         ServerExplosionInfo.HitNormal = HitNormal;
         ForceNetUpdate();
 
         SetActorEnableCollision(false);
         ProjectileMovementComponent->StopSimulating(FHitResult());
-        //SetActorHiddenInGame(true);
-        SetLifeSpan(0.2f);
+
+        // Give 200ms to replicate explosion before destroy
+        SetLifeSpan(0.200f);
```
Comment (wail) [inline Source/OpenTournament/UR_Projectile.cpp:275]: Magic number here. Should we parameterize this?

Comment (Saibamen) [reply Source/OpenTournament/UR_Projectile.cpp:275]: IMHO yes. Will be helpful for mutators/mods etc

File: Source/OpenTournament/UR_Projectile.cpp [hunk 2]
```diff
@@ -209,24 +210,80 @@ void AUR_Projectile::DealSplashDamage()
...
 
+    if (bNetTemporary || GetNetMode() == NM_Standalone)
+    {
+        Destroy();
+        return;
+    }
+
     if (HasAuthority())
     {
+        //UKismetSystemLibrary::PrintString(this, TEXT("HasAuthority > ServerExplosionInfo"));
```
Comment (Saibamen) [inline Source/OpenTournament/UR_Projectile.cpp:265]: Cleanup after yourself

File: Source/OpenTournament/UR_Projectile.cpp [hunk 3]
```diff
@@ -209,24 +210,80 @@ void AUR_Projectile::DealSplashDamage()
...
+*     (note: client cannot destroy as long as replication link exists)
+*
+* (4) Receive ServerExplosionInfo after exploded on client.
+*     > See case (1), this is either a confirmation, or an unreg case.
+*     > In unreg case, we should play the real explosion now.
+*/
+
+void AUR_Projectile::OnRep_ServerExplosionInfo()
+{
+    //UKismetSystemLibrary::PrintString(this, TEXT("OnRep_ServerExplosionInfo"));
```
Comment (Saibamen) [inline Source/OpenTournament/UR_Projectile.cpp:236]: Cleanup after yourself

File: Source/OpenTournament/UR_Projectile.cpp [hunk 4]
```diff
@@ -209,24 +210,80 @@ void AUR_Projectile::DealSplashDamage()
...
+
+    // Either way, stop simulating and wait for server destruction (client cannot destroy networked actor).
+    SetActorHiddenInGame(true);
+    SetActorEnableCollision(false);
+    ProjectileMovementComponent->StopSimulating(FHitResult());
+}
+
 void AUR_Projectile::Explode(const FVector& HitLocation, const FVector& HitNormal)
 {
+    //UKismetSystemLibrary::PrintString(this, TEXT("Explode"));
```
Comment (Saibamen) [inline Source/OpenTournament/UR_Projectile.cpp:253]: Cleanup after yourself

File: Source/OpenTournament/UR_Projectile.cpp [hunk 5]
```diff
@@ -209,24 +210,80 @@ void AUR_Projectile::DealSplashDamage()
...
-        //SetActorHiddenInGame(true);
-        SetLifeSpan(0.2f);
+
+        // Give 200ms to replicate explosion before destroy
+        SetLifeSpan(0.200f);
     }
     else
     {
-        Destroy();
+        //UKismetSystemLibrary::PrintString(this, TEXT("Client > Hide"));
```
Comment (Saibamen) [inline Source/OpenTournament/UR_Projectile.cpp:279]: Cleanup after yourself

**Pull request merged**